### DISCUSSION TITLE
Is there a way to make an additional float[] output showing individual phoneme times/lengths from the native DLL?

### DISCUSSION THREAD

**Discussion by JasonBlain** [rhasspy/piper#425] [Q&A] [answered]
I'm over in Unity and trying to lipsync visemes to the audio output, and I can see where the function to generate the audio from the phonemized text is appending the bits of phonemes together, but the final output length of the audio clip is not the whole story in terms of phoneme/viseme pacing.

Is there a way to add an output array of floats to go with the Phoneme sequence of a TTS output? No need to edit the tensor, just need an additional output with the array after it has been rendered. I imagine it would have to be done at the .cpp level and then rebuilt into the DLL?

**Comment by synesthesiam** [comment]
Yes, this is fully possible but will require either (1) using the original PyTorch models or (2) re-exporting the voice models and changing the C++ code.

An intermediary product of the model is the length of every phoneme. This can be returned with the audio, but will require the changes above.

**Comment by JasonBlain** [comment]
Hey thanks for the reply.

How could I re-export with the additional outputs? I've got some python experience, kind of a generalist. I've got most of a modified overload header written in the cpp files of a fork of the piper.unity repo I made, so I have that part.. worked out, I think.

**Reply by synesthesiam** [reply, marked as answer]
The `w_ceil` variable has the phoneme lengths: https://github.com/rhasspy/piper/blob/e5cb84cd5c3cb93eff9398264d5a7da5122ff21a/src/python/piper_train/vits/models.py#L703
Multiplying this tensor by 256 will get you the number of audio samples per phoneme.

That `w_ceil` tensor needs to be returned from the `infer` function and then also returned with the audio here: https://github.com/rhasspy/piper/blob/e5cb84cd5c3cb93eff9398264d5a7da5122ff21a/src/python/piper_train/export_onnx.py#L60

On the C++ side, you then need to pick apart the multiple output tensors (one audio, one phoneme samples): https://github.com/rhasspy/piper/blob/e5cb84cd5c3cb93eff9398264d5a7da5122ff21a/src/cpp/piper.cpp#L386

**Reply by JasonBlain** [reply]
Lovely, will go through that as best I can and see what I can come up with. Good cites. Thank you.

**Reply by JasonBlain** [reply]
EDIT: Found checkpoints repo on HuggingFace and will try re-export.

**Reply by JasonBlain** [reply]
All of those citations were great, I did have to do a couple more things for my system config to be able to push the new ONNX export out from python.

I'm on windows so I had to do some edits to export_onnx.py, do some POSIX path Windows-explicit things. Monotonic align was acting up and had a circular reference so I had to change the setup and init on it just a bit, mostly path related issues.

**Reply by adityapandey9** [reply]
Hi @synesthesiam Can I please know why you multiplied it with 256? Is it because the number of the number of symbols mentioned in the `config.json`? Let me know if I am wrong.

**Comment by JarbasAl** [comment]
Is it possible to expose this in Piper python api? It would help a lot in OVOS to generate mouth movements for the Mark1 device

together with the generated audio file in OVOS we need a list of phonemes + duration of each phoneme, for the most part we just use the original mimic1 TTS to generate these, but this is far from perfect as they often dont match the actual audio, if we could get piper to output this info natively it would be awesome!

**Reply by synesthesiam** [reply]
Yes, I will expose this to both the C and Python APIs in the future.

**Reply by goldyfruit** [reply]
Would help a lot indeed as for now we have to install Mimic TTS which is around 400Mb :+1:

**Reply by goldyfruit** [reply]
Here is an example when using Piper with `ryan-low` voice on a Mark 1 device.

https://youtube.com/shorts/9m3501afUEw

**Reply by MetaMachina** [reply]
> Yes, I will expose this to both the C and Python APIs in the future.

Hi Michael ( @synesthesiam ), is it possible to share an approximate time-frame as to when the next release of Piper, with the capability to export phonemes and their timestamps to a file, will be published? Thanks you.
